File: packages/imanikurd-prayer/src/qibla.ts
import type { Coordinates, QiblaResult } from "./types.js";

/** Kaaba coordinates in Mecca */
export const KAABA_COORDINATES: Coordinates = {
  lat: 21.4225,
  lng: 39.8262,
};

/**
 * Calculate Qibla direction (bearing in degrees from North, 0-360)
 * from any latitude/longitude to the Kaaba.
 */
export function calculateQibla(latitude: number, longitude: number): number {
  const phiK = (KAABA_COORDINATES.lat * Math.PI) / 180;
  const lambdaK = (KAABA_COORDINATES.lng * Math.PI) / 180;
  const phi = (latitude * Math.PI) / 180;
  const lambda = (longitude * Math.PI) / 180;

  const psi =
    (180 / Math.PI) *
    Math.atan2(
      Math.sin(lambdaK - lambda),
      Math.cos(phi) * Math.tan(phiK) - Math.sin(phi) * Math.cos(lambdaK - lambda)
    );

  return Math.round((psi + 360) % 360);
}

/** Get full Qibla result with direction and coordinates */
export function getQibla(from: Coordinates): QiblaResult {
  return {
    direction: calculateQibla(from.lat, from.lng),
    kaaba: KAABA_COORDINATES,
    from,
  };
}

/** Check if a compass heading is pointing toward Qibla (within tolerance degrees) */
export function isFacingQibla(
  heading: number,
  latitude: number,
  longitude: number,
  tolerance = 10
): boolean {
  const qibla = calculateQibla(latitude, longitude);
  const diff = Math.abs(((heading - qibla + 180) % 360) - 180);
  return diff <= tolerance;
}

/** Get distance to Kaaba in kilometers (Haversine formula) */
export function getDistanceToKaaba(latitude: number, longitude: number): number {
  const R = 6371;
  const dLat = ((KAABA_COORDINATES.lat - latitude) * Math.PI) / 180;
  const dLng = ((KAABA_COORDINATES.lng - longitude) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((latitude * Math.PI) / 180) *
      Math.cos((KAABA_COORDINATES.lat * Math.PI) / 180) *
      Math.sin(dLng / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return Math.round(R * c);
}

/** Get Kaaba coordinates */
export function getKaabaCoordinates(): Coordinates {
  return KAABA_COORDINATES;
}

File: packages/imanikurd_flutter/lib/src/qibla.dart
import 'dart:math' as math;

import 'models.dart';

const Coordinates kaabaCoordinates = Coordinates(lat: 21.4225, lng: 39.8262);

int calculateQibla(double latitude, double longitude) {
  final phiK = kaabaCoordinates.lat * math.pi / 180;
  final lambdaK = kaabaCoordinates.lng * math.pi / 180;
  final phi = latitude * math.pi / 180;
  final lambda = longitude * math.pi / 180;

  final psi = 180 / math.pi *
      math.atan2(
        math.sin(lambdaK - lambda),
        math.cos(phi) * math.tan(phiK) - math.sin(phi) * math.cos(lambdaK - lambda),
      );

  return ((psi + 360) % 360).round();
}

QiblaResult getQibla(Coordinates from) => QiblaResult(
      direction: calculateQibla(from.lat, from.lng),
      kaaba: kaabaCoordinates,
      from: from,
    );

bool isFacingQibla(
  double heading,
  double latitude,
  double longitude, {
  double tolerance = 10,
}) {
  final qibla = calculateQibla(latitude, longitude);
  final diff = ((heading - qibla + 180) % 360) - 180;
  return diff.abs() <= tolerance;
}

int getDistanceToKaaba(double latitude, double longitude) {
  const r = 6371.0;
  final dLat = (kaabaCoordinates.lat - latitude) * math.pi / 180;
  final dLng = (kaabaCoordinates.lng - longitude) * math.pi / 180;
  final a = math.pow(math.sin(dLat / 2), 2) +
      math.cos(latitude * math.pi / 180) *
          math.cos(kaabaCoordinates.lat * math.pi / 180) *
          math.pow(math.sin(dLng / 2), 2);
  final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
  return (r * c).round();
}

Coordinates getKaabaCoordinates() => kaabaCoordinates;
